### ISSUE TITLE
Watch does not automatically reconnect to GadgetBridge after turning airplane mode on and off

### ISSUE THREAD

**Issue by GithubUser5462** [InfiniTimeOrg/InfiniTime#302]
InfiniTime version: 1.0.0 RC3
Gadgetbridge version: 0.56.1

Steps to reproduce:

1. Pair InfiniTime with Gadgetbridge, enable the "Reconnect automatically" setting in GB
2. While the two are connected turn  airplane mode on and then off after the watch is disconnected
3. GB tries to reconnect, first with status "Connecting" and then gets stuck with "Waiting for reconnect"

The workaround I have found is to fully disconnect in GB - status "Not connected" and then pair the watch again.
Restarting the watch or Gadgetbridge did not seem to help.

I don't know if the problem is with GB or IT but it would be nice if the two could automatically reconnect in this scenario if possible.

**Comment by petterhs** [comment]
hmm I'm not able to reproduce this issue running same versions. Do you experience the same if you toggle Bluetooth  off/on on the phone?

**Comment by GithubUser5462** [comment]
Yes, exact same result.

**Comment by JF002** [comment]
This kind of issue is really hard to debug, and is probably caused by a mix of PineTime HW, InfiniTime, phone HW and software stack. 
InfiniTime does advertise itself automatically as soon as it's disconnected during 3 minutes. It also restarts advertisting every time the display is switched on. Then, it's up to the companion app to receive these advertising messages and reconnect to the watch.

**Comment by pfeerick** [comment]
Not knowing that behaviour is most likely the 'cause' of the problem... i.e. I would probably have had a similar complaint if I didn't know that turning the screen on wakes up advertising. 

> InfiniTime does advertise itself automatically as soon as it's disconnected during 3 minutes. It also restarts advertisting every time the display is switched on. Then, it's up to the companion app to receive these advertising messages and reconnect to the watch.

Maybe a future thought could be have periodic advertising... say once every five minutes or so, for 10 seconds or whatever (I'm sure there's some convention that can be followed - or not!) ... i.e. try and make itself available of the phone wants to reconnect.

**Comment by GithubUser5462** [comment]
I made a couple more tries to try to understand what is happening.
It seems that it is not limited to Gadgetbridge since it also happens with nRF Connect.

In nRF Connect when I disable/enable bt the "bonded" tab in nRF Connect gets wiped, but the scanner does detect watch - as not bonded.
If i try to manually connect sometimes it connects, but most of the time it timeouts/errors - logger says `Error 133 (0x85) GATT ERROR`. Which is probably not very helpful.

**Comment by pfeerick** [comment]
Whether the device is bonded or not has no impact on the device appearing
in the scanner tab. It is merely a list of all the devices which are
currently averrising, with greyed out entries for those devices that have
recently advtertise, but not presently advertising (Bluetooth advertising
of presence).

Can you try this... Start a scan (in NRF Connect), wait for a few seconds
and confirm the InfiniTime is not active. Then whilst NRF Connect is still
scanning,  press the button on the pinetime momentarily so the screen comes
on. Does it show up then? If so, it is working as intended.

**Comment by diegosolor** [comment]
I have the same unstable connection problem with gadgetbridge, only workaround is unpair/repair the pinetime

**Comment by jjrugui** [comment]
I can confirm the same issue with reconnection as @diegosolor . Connections gets lost and some times pinetime does not appear in gadgetbridge. Only solution in that case is to restart the watch.

For reference I'm using a Oneplus Nord with A droid 11. I'm a bit lost with how to debug it but let me know if I can do anything to provide help/info about the reconnection issue.

**Comment by pfeerick** [comment]
Sometimes I have to tell gadgetbridge to connect twice... I.e. connect, press and hold to disconnect after a few seconds of no activity, and then connect again. With the pinetime screen on to ensure it is actually advertising.

**Comment by blaueente** [comment]
Independent of airplane mode: Mine (1.0.0) disconnects sometimes from gadgetbridge, and then does not appear on the nRF scanner, and then can only be reconnected by restarting the pinetime.

**Comment by jjrugui** [comment]
I'm having the exact same behaviour as @blaueente . I get multiple disconnects per day and I need to restart the pinetime in order to be able to pair it again.

**Comment by Anyil89** [comment]
I am on arch Linux arm danct12s  Image. I’m using it on my pinephone,  when I restart my phone like when I update the Bluetooth says it’s disconnected.  But it was connected before the reboot. (my PineTime) But sometimes rarely it will reconnect itself after a while but 80% of the time I have to restart my watch to make it reconnect or disconnect from amazfish and then reset that back up. For it to register.

**Comment by jjrugui** [comment]
Today I upgraded Gadgetbridge to latest version 0.58.0 and so far I had a way more stable connection and automatic reconnect working. I'll give more info after testing it for a couple days but I highly recommend the upgrade.

**Comment by NCLI** [comment]
I pretty much have to restart the PineTime every 20 hours to connect. Not very useful like that.

**Comment by 1989gironimo** [comment]
I can add, that not only activating airplane mode causes this behavior. Disabling bluetooth for a while triggers this issue as well, even on the newest firmware (1.2 at this time) and newest gadgetbridge (0.58.1) on GrapheneOS (RQ3A.210705.001.2021.07.07.19)

I even have to remove it from gadgetbridge first, restart the phone AND the Pinetime to get a working connection, so this might be somewhat related, but not the same issue..

**Comment by raananelefant** [comment]
I can report that upon Airplane mode on/off, PineTime, running Infinitime FW v1.1.0 w/Hebrew support, cannot reconnect automatically/manually using latest Gadgetbridge (v0.58.2).

Workaround is "disconnecting" watch via GB, re-pairing via Android settings (PineTime disappears from "Previously connected devices") and then connecting via GB.

Phone is a Planet Computer Cosmo Communicator running Android v9, Build number Cosmo-9.0-Planet-02032021-V25, Custom build version alps-mpp-p0.mp1-V5.313

**Comment by Roterfux** [comment]
Same issue with firmware 1.3.0

**Comment by aemkai** [comment]
I have the same problem. 
Watch doesn't reconnect to Gadgetbridge or nRF Connect and also could not be found by BL-Scan of phone. BL-Symbol on watch is off.
Cycle Bluetooth connection or reset of phone doesn't help.
After reset of watch connection is established automatically and without problems with both apps.
Issue appears @ about 20 hours runtime.
Firmware tested with V1.2.0, self compiled V.1.2.0 and self compiled V1.3.0

**Comment by tealeg** [comment]
Another one here.  I just received two PineTime's, the current firmware is v1.2.0.  The behaviour is identical on both devices, and with two different Android phones running Gadgetbridge. 

My experience is as per #527 - there's no need for airplane mode, or any shenanigans on the phone, at some point, when the watch is idle, it will disconnect and will no longer advertise until a restart is forced.   As others have noted, the Bluetooth indication on the watch face disappears, and (naively) I assume the Bluetooth radio has been powered down.

**Comment by jamesstott** [comment]
Same problem here as well with v1.3.0 an Gadgetbridge.

**Comment by 1989gironimo** [comment]
> Another one here. I just received two PineTime's, the current firmware is v1.2.0. The behaviour is identical on both devices, and with two different Android phones running Gadgetbridge.
> 
> My experience is as per #527 - there's no need for airplane mode, or any shenanigans on the phone, at some point, when the watch is idle, it will disconnect and will no longer advertise until a restart is forced. As others have noted, the Bluetooth indication on the watch face disappears, and (naively) I assume the Bluetooth radio has been powered down.

Well, in my case (I've tested it already several times) it is NOT sufficient just to restart the watch.
I've to switch off bluetooth and switch it on again (instead of rebooting, that's a little faster now :) ), because Gadgetbridge cannot scan for devices correctly. If I don't do this, it just hangs and gets unusable.

But the watch also gets disconnected after some time, as someone described above and doesn't reconnect anymore after approx. 20h.

**Comment by JackofSpades707** [comment]
I just wanted to chime in and say, if I reboot my phone, it won't connect to my pinetime without unpairing and repairing with GadgetBridge. (stuck on waiting for reconnect)
I suspect the same will be true if I toggle airplane mode or disable/enable Bluetooth.

PineTime: 
- v1.2.0

Phone:
- OnePlus 6T (Sprint model) - A6013
- OxygenOS 10.3.10 (rooted, no modules mucking with bluetooth)

GadgetBridge: 
- version: 0.58.2
- CompanionDevice pairing: enabled

**Comment by bearclaws8** [comment]
I can confirm I have the same problem with a new PineTime. Just like @heinrich-ulbricht I received watch, updated firmware from 1.2.0 to 1.3.0, paired with Gadgetbridge, disabled Bluetooth, enabled Bluetooth... no connection anymore.

I have had some better connectivity with nRF Connect compared to GB.  GB seems to dislike reconnecting when toggling Bluetooth while nRF will usually reconnect.  However, after a while, like noted in issue #527 the PineTime itself will no longer broadcast via BLE.

Might it be that this issue and #527 are connected but due to two different issues, one that GB is a flaky and the other where PineTime stops broadcasting?

Is it possible to add into the settings the ability to turn Bluetooth on/off and force the service to restart without restarting the entire device?  On/off may be nice for those looking to maximize battery and security and restarting the BLE broadcast may help with debugging.

**Comment by Avamander** [comment]
> GB seems to dislike reconnecting

It doesn't if the disconnection is too long, primarily because making a new connection itself is very unreliable right now. Once InfiniTime can keep a connection in-range and has working reconnects, Gadgetbridge's auto-reconnection can be dialed up.

**Comment by bearclaws8** [comment]
> It doesn't if the disconnection is too long, primarily because making a new connection itself is very unreliable right now. Once InfiniTime can keep a connection in-range and has working reconnects, Gadgetbridge's auto-reconnection can be dialed up.

@Avamander  thank you for the clarification.

**Comment by heinrich-ulbricht** [comment]
This issue seems to be a rather complex one due to the number of potentially involved parts (and parties):

- Gadgetbridge might have an issue with reconnecting
- Infinitime might have an issue with reconnecting
- the watches hardware might have an issue with reconnecting (?)
- maybe the phone or its operating system plays a role, too?

Furthermore only some users seem affected. (Otherwise this would have priority over new features? Can we estimate how many users are affected?) So the issue might prove hard to reproduce for those who are capable of analyzing and/or fixing this issue.

From an end user's point of view it's frustrating. The connection between watch and phone is unusable for me. Going out of range often is enough to kill the Bluetooth connection. I then have to re-pair, which takes time and after some time I don't bother anymore. I also see that this issue has been known a long time already which gives me the fear it will never be fixed.

From a developer's point of view I'd like to help get this sorted out. I just don't know how and somebody would have to tell me what to do to diagnose this. Reading logs with adb? Can do. Fire up a dev environment? Not so sure. What else?

How is the current state for this issue? Is somebody currently working on it? Can we assume that this will sort itself out with future updates to software and/or hardware?

Back in the Windows Phone times I used to write: I really want this to work. Seems like I'm back again writing: I really, really, reeeeeally want this to work!!

**Comment by Cuperino** [comment]
My PineTime is having similar issues. It won't reconnect after a few minutes. Nevertheless, existing connections can last for hours. I have not yet checked what's the longest it will last.

Upon resetting I sync to both my laptop, using Siglo, and my stock Android phone, using GadgetBridge, one after the other before the PineTime's screen turns off. After disconnecting, the paired device needs to be removed from GadgetBridge and the PineTime needs to be reset for it to connect to either device.

PineTime:

    FW: v1.3.0 and v1.2.0
    HW: v1.0.0

Phone:

    Google Pixel 3a
    Android 11 (stock)

GadgetBridge:

    version: 0.59.0
    CompanionDevice pairing: enabled

Laptop:

    Dell XPS 13 7390 2-in-1
    KDE neon Testing Edition
    Linux Kernel 5.11.0-27-lowlatency

Siglo:

    version: Flathub stable released on August, 04 2021

**Comment by Debo-rah** [comment]
The same here. Sometimes it helped to reinstall gadgetbridge, sometimes it helped to reboot the watch. After updating to firmware v.1.3 it seemed a little better, but then I got the same problem as someone mentioned above, that it got stuck at the infinitime logo, even though I DID validate the new fimware.

`
FW: v1.3.0 and v.1.2.0
HW: v1.0.0`

Phone:
`Google Pixel 2
Android 10 (Lineage 17.1)`

GadgetBridge:
`version: 0.59.0`

**Comment by heinrich-ulbricht** [comment]
@aemkai I read this, too, but this seems like a second problem. For me I 100% can reproduce non-connectivity by disabling+enabling my phone's Bluetooth adapter. No connection afterwards. Fixed by removing+re-adding the watch from Gadgetbridge.

As for reconnecting after being out of range: this seems to work. I fooled myself once or twice when I disabled location services, then went out of Bluetooth range and wondered why it didn't reconnect after coming into range again. Seems like location needs to be enabled for this to succeed.

**Comment by aemkai** [comment]
@heinrich-ulbricht 
i can disable and re-enable without problems. The bluetooth icon at the watch also disappears and vice versa.
Location is disabled most time without problems.
Maybe the issue isn't caused by the watch and should be searched at the app in combination with the OS.

**Comment by Avamander** [comment]
@aemkai If you look at this current thread then you can see how the issue isn't OS or companion-specific.

**Comment by aemkai** [comment]
@Avamander 
Why?
Some users say, a restart of watch isn't sufficient. Others must remove the watch from Gadgetbridge and add it again. This doesn't seem to be an issue of the watch (alone).
The next point is, why some users have problems, others not. An explanation could be the difference between pre- and self-compiled firmware, but this doesn't mitigate the other points.

**Comment by Avamander** [comment]
@aemkai I'm not sure what's unclear? The issue exists across multiple OSs and companions (which can also handle other watches great), it's unlikely to be specific to those.

**Comment by aemkai** [comment]
@Avamander This doesn't explain my considerations. Also I wrote "app in combination with the OS". Not every app has the same behaviour even on different builds of the same OS. Then e.q. energy saving or privacy settings have different behaviours.
If OS/companion isn't the issue, why isn't a reset of the watch the solution? 
For example @Debo-rah wrote, she had to reinstall Gadgetbridge. How should this need be caused by the watch?

**Comment by Avamander** [comment]
@aemkai
> why isn't a reset of the watch the solution?

Because Bluetooth stacks can make assumptions when bonded, that InfiniTime violates.

> How should this need be caused by the watch?

It wasn't, the reinstall was very likely unnecessary.

**Comment by aemkai** [comment]
@Avamander I played arond a little bit with the 2 watches, changed airplane mode, bluetooth dis-/enable, (dis-)bonding etc. and got it (I dont know how exactly) to an point, where a connection to both watches wasn't possible, neither with Gadgetbridge nor with nRF Connect.
I reset one watch, no improvement. I reset the phone, and afterwards both watches (also the non-reset) connected without problems.
Maybe there is an issue at Pinetime - i won't exclude that - but i think, there is also a problem in interaction of different components.

**Comment by tmilburn** [comment]
I get the feeling there is an integer overflow bug somewhere. At a 32768Hz clock rate a 32 bit signed integer overflows after just over 18 hours. I have definitely seen the Bluetooth inactive when checked at 19h30m of uptime and other commenters have mentioned 20 hours.

**Comment by bearclaws8** [comment]
@tmilburn could an integer overflow cause other devices to refuse to see/connect to the watch after it is reset?  I have noticed that after a few instances of PT dropping Bluetooth, and me restarting the watch, Android will begin to not even see it via Bluetooth without the phone being restarted.  Neither OS Bluetooth settings, Gadgetbridge, nor nRF Connect will pick it up.  Granted this is likely an Android issue/limitation but wasn't sure if it was a behavior that may indicate something the watch was doing.

Disregard if this is not helpful.

**Comment by heinrich-ulbricht** [comment]
Today I may have experienced a combination of both issues. The watch did not connect. I removed it from Gadgetbridge and tried to add it again. Usually this helps. But not this time as I also had to reset the watch for Gadgetbridge to find it. This might be the ~20 hour issue.

I wonder how this issues can be tackled. Solving it needs a combination of people able to 

1. reproduce the issue(s), 
2. provide data and logs to analyze, 
3. locate the error(s) and 
4. finally fix them.

We definitely have folks here who can reproduce the issues. I also suspect we've got folks to fix errors. But points 2 and 3 might be difficult. Does somebody know which logs we could check to get useful hints? I'll try an `adb logcat` the coming week. Who knows if something comes up. Unfortunately I have no dev kit at hand so I cannot look into watch logs (I suppose).

**Comment by bryantqo** [comment]
Happening daily for me as well. Restart watch, delete from GB, re-pair

**Comment by tmilburn** [comment]
@heinrich-ulbricht 

1. Reproducing the issue is relatively easy, keep the watching running for 20 hours. There could be a way to speed this up by starting timers with an initial value close to the signed int rollover.
2. Android logs might be helpful in showing if the issue was whilst connected or whilst advertising
3. I have been looking through the bluetooth stack at all the timers to see if there are any issues with them and so far no luck in finding the error.

**Comment by tealeg** [comment]
20 hours is a read herring or a different issue, mine does it sometimes in
less than one hour.

**Comment by pbryan** [comment]
I don't think it's a red herring. When phone and watch are left in
proximity, I see ~20 hours consistently.

**Comment by bryantqo** [comment]
Could it not be both? Regardless there is a disconnect and it doesn't want to reconnect. Another thing I can test is going out of range to see if it reconnects.

**Comment by bearclaws8** [comment]
I can confirm that on my setup it will reconnect after separation. I am frequently away from my phone for several hours and my watch reconnects without issue (within the 20 hours).  However, I have had difficulty with multiple reconnects.

To me it seems the duration of disconnect due to separation does not matter (within the 20 hours) but disconnecting too frequently has sometimes shortened the "clock" for the watch not reconnecting. This usually happens when I'm running around so I haven't been able to pay enough attention to reproduce reliably.

I used to have more frequent issues like @tealeg but haven't had that happen much lately. Usually the shorter time for not reconnecting was due to me coming into and out of range very frequently over that timeframe.

**Comment by tmilburn** [comment]
I managed to run a test to confirm that one of the issues is that the watch does not advertise when the uptime has surpassed 18 hours 12 minutes (65536 seconds or 2147483648 (2^31) ticks). I connected to the watch at 18 hours successfully and checked that the connection was still up at 18:13. At this point I manually disconnected and attempted to reconnect. This failed and I subsequently couldn't detect any advertisement packets from the watch. 

There must be an integer overflow somewhere causing this problem but I haven't managed to track it down yet. Nor do I have an open unit which I can attach a debugger to.

**Comment by pbryan** [comment]
Nice detective work!

**Comment by aemkai** [comment]
> I managed to run a test to confirm that one of the issues is that the watch does not advertise when the uptime has surpassed 18 hours 12 minutes (65536 seconds or 2147483648 (2^31) ticks).

How did you calcualte the 2^31 ticks?
The tick frequency should be 1024 Hz (see DateTimeController.cpp and FreeRTOSConfig.h), so 65536 seconds are 67.108.864 ticks (= 2^26). Thus, in my eyes an overflow of a 16-bit-variable should be more likely

**Comment by bellegarde-c** [comment]
Found this in my LineageOS logcat logs:

```
09-01 18:00:45.854  2467  3171 E bt_btif : bte_scan_filt_param_cfg_evt, 23
09-01 18:01:04.123  2467  3171 E bt_btm  : Ignoring RNR as the state is not BTM_SEC_STATE_GETTING_NAME
09-01 18:01:09.286  2467  3171 E bt_btm  : Ignoring RNR as the state is not BTM_SEC_STATE_GETTING_NAME
09-01 18:01:09.289  6476  6476 E nodomain.freeyourgadget.gadgetbridge.activities.DiscoveryActivity: Problem with permissions, returning

```

**Comment by lgs11** [comment]
PineTime 1.3.0
Gadgetbridge 0.59.2
/e/OS Q-0.18 (based on LineageOS)

Mine disconnects any time the phone is too far from the watch and doesn't reconnect again.

What works for me is to press "Connect new device," select the watch, and then press "Do not pair."

**Comment by lgs11** [comment]
Maybe it's just me, but reconnecting seems to work much better now after upgrading GadgetBridge to 0.59.3. Most of the time, I just have to wake the watch to cause a reconnect.

**Comment by bellegarde-c** [comment]
Works better now except if I put my phone in airplane mode.

**Comment by bryantqo** [comment]
Updated to v1.4.0. At some point overnight it lost connection and is no longer advertising. Using nRF Connect the device is not advertising and cannot connect if previously bonded.

**Comment by M4xTheMan** [comment]
I'm not able to connect to my PineTime using Gadgetbridge after I've turned off/on bluetooth or disable airplane mode.

I've to delete the device in gadget bridge and pair it again. Therefore I've to do it everday...

phone OS: Lineage OS 18.1
Android version: 11
Gadgetbridge: 0.59.3
Infinitime: 1.4.0

**Comment by ainola** [comment]
Looking through the thread shows nearly all users on LineageOS. I just to confirm that an xperia xz2 with (woefully out of date) stock firmware on Android 10 has this exact issue.

**Comment by tealeg** [comment]
I'm using a Xiaomi Redmi Note 7 with MIUI 12.03 on Androd 10
QKQ`.190910.002 - if that's useful to anyone.  Clearly not a LineageOS
specific issue.

**Comment by akutuzov** [comment]
I am experiencing the same on Sony Xperia 10 with Sailfish OS 4.2.
It does not have anything to do with applications (I am using AmazFish): the phone itself at some point stops re-connecting to the watch. As previous commentators mentioned, it happens after random time intervals, usually an hour or so, but sometimes a day.

Interestingly, the phone sees the watch via Bluetooth (so it _is_ advertising), but upon an attempt to re-connect, it fails with "Problem with pairing authentication. The passcodes did not match". At this point, only resetting the watch helps.

**Comment by grassygnoll** [comment]
New PineTime(r) - Just chiming in with another data point; Pixel 3a (_Android 11_); Gadgetbridge 0.59.3; and sealed PineTime (_InfiniTime 1.4; Build date Sep. 2, 2021 18:36:27_).

Received my watch a few days ago and installed necessary app on phone, then Paired/Bonded and was good to go for about 7-8 hours and then noticed Gadgetbridge no longer was connected to watch.  Have not been able to reconnect since (_have yet to restart/reset the watch but did do all the toggling on the app, Bluetooth, phone restart, etc.  Removed the watch from the Gadgetbridge list of devices and it cannot find it anymore to pair with.  Will restart the watch and (_based on this issue log_) suspect it will reconnect for a short while.

**Comment by danielgjackson** [comment]
(Apologies that there's a lot of text -- but hopefully it can lead someone in the right direction.)

I am also seeing the advertising consistently stop after 18.2 hours.  The application firmware continues as normal: successfully requests to start advertising for 2000ms, is notified by the BLE stack that it has stopped after 2000ms, and it starts it again -- no errors seem to surface to the application.  

For reference, `NRF_RTC0` *is* at 32768 Hz, and it is the timer used by *mynewt* (`OS_CPUTIME_TIMER_NUM=5`).  The `systickCounter` uses `NRF_RTC1` with a prescaler to bring it down to 1024 Hz.

The `NRF_RTC0` timer is used to provide `os_cputime_get32()` (through `hal_timer_read()` then `hal_timer_read_bsptimer()`).  It is also used in calculations to set a compare register used to fire an interrupt at a specific time in the future.

The timer is 24-bits and wraps every 512 seconds.  A 32-bit timer is required at a higher level.  For this, the top 8-bits are kept in a variable as the top byte of a 32-bit value (OR'd with the counter when the time is read) and incremented on overflow -- this is referred to as `bsptimer->tmr_cntr`.

When I experimentally initialize this top byte to 0x7F, my device stops advertising after just one 8m32s (512 second) overflow period.  You should be able to reproduce this by modifying `src/libs/mynewt-nimble/porting/nimble/src/hal_timer.c`, at line 472, after `bsptimer->tmr_rtc = 1;`, insert:

```c
bsptimer->tmr_cntr = (uint_32)0x7f << 24;  // HACK: Initialize the hal_timer to overflow, when treated as a SIGNED 32-bit value, after 512 seconds (8m32s).
```

In addition, when I initialize this top byte to 0x7E (one value less), my device stops advertising after twice the overflow period (17m04s, or 1024 seconds).  

Something about this rollover (into the top bit) seems to be causing the advertising to stop.  Importantly: the full 32-bits are not overflowing, but this value would now be negative if it was ever treated as a signed 32-bit value.  I don't believe it is directly, but it may well be treated as signed when a delta value is computed as the difference between two times, and perhaps some logic involving this is incorrect.  

In the interrupt handler `hal_rtc_timer_irq_handler()` (via `nrf52_timer5_irq_handler()`, via `rtc0_isr_addr`, via `isr_rtc0()`), typically a few *compare* events happen every second, and the *overflow* happens every 512 seconds.  After the overflows cause `tmr_rtc` to rollover into the top bit, the compare events suddenly stop.  The timer is left running, and the overflow events still happen. 

The reason the compare events stop firing is that the compare value `rtctimer->EVENTS_COMPARE[NRF_RTC_TIMER_CC_INT]` seems to be around 0x00800000 away (modulo 0x00ffffff) from the current time values.  Such a value is set in `nrf_timer_set_ocmp()` line 178: the compare value to the current counts plus (1<<23) -- but only if the `delta_t` >= (1<<24).  (The `RTC0` compare register itself seems to be -- sensibly -- only 24-bit, and this value isn't masked.)

`nrf_timer_set_ocmp()` is called to deal with the `hal_timer_q` queue entries -- populated from `hal_timer_start_at()` via `os_cputime_timer_start()`, and that seems to be mainly called by `ble_ll_sched_*()` functions -- e.g. `ble_ll_sched_adv_new()`.  

So it is presumably that these compare events are no longer firing (seemingly because of their incorrect `delta_t` values) which affects the logic in the link-layer and prevents the advertising packets from being sent properly.

However, I can't yet see how/why the `delta_t` is negative/large -- I don't have my board soldered up to the debugger, and I can't keep a temporary connection for multiple minutes!  If this is a problem in the stack (and not something in the application/config), I wonder why this wouldn't have been noticed upstream on NimBLE before (although I notice the in-tree version of NimBLE is not that current).

**Comment by M4xTheMan** [comment]
Could it be that we're seeing two errors here?

1. Pinetime stops advertising after x hours and has to be restarted
2. Gadgetbridge is not able to reconnect with Pinetime, the pairing has to be removed and the watch has to be paired again

**Comment by Avamander** [comment]
@M4xTheMan No, but yes, but actually no. GB can't reconnect if InfiniTime doesn't advertise or if it has mucked up Android's Bluetooth stack somehow (which might be different from the advertising issue, but that's unknown).

**Comment by Avamander** [comment]
@danielgjackson 

>  I wonder why this wouldn't have been noticed upstream on NimBLE before (although I notice the in-tree version of NimBLE is not that current).

Updating it to latest did not remedy the issue, has been tried.

**Comment by tealeg** [comment]
This is definitely two different bugs. I've never gotten beyond a few hours
without hitting the problem, 18 hours would be a dream.

**Comment by Avamander** [comment]
> This is definitely two different bugs.

That is yet to be determined.

**Comment by danielgjackson** [comment]
I think I've finally fixed the problem with the advertising packets stopping after 18.2 hours.  I can't say for sure how much this might account for the various connectivity problems described here and in other issues -- but it should hopefully be some improvement.

It seems to have been caused by the function `ble_ll_rfmgmt_enable_now()` (`ble_ll_rfmgmt.h`).  (It's implemented in the header as a `static inline` because `MYNEWT_VAL_BLE_LL_RFMGMT_ENABLE_TIME` is not defined in `syscfg.h`.)

The function "Enables RF immediately and returns tick at which RF will be fully enabled", however the placeholder function actually always returns `0`:

```c
static inline uint32_t ble_ll_rfmgmt_enable_now(void) { return 0; }
```

This causes an issue in `ble_ll_adv_sm_start()` (`ble_ll_adv.c`) -- it retrieves this time and a delta is calculated to check whether the advertising is trying to start before the RF would be ready:

```c
earliest_start_time = ble_ll_rfmgmt_enable_now();
// [...]
delta = (int32_t)(advsm->adv_sch.start_time - earliest_start_time);
if (delta < 0) {
// [...]
```

As this is not a tick time near to the current value (but instead, always 0), this calculation will overflow once the timer gets to 0x80000000, causing an incorrect 9 hour forward adjustment to be applied to the schedule.  This later causes `nrf_timer_set_ocmp()` to treat the value as "too far ahead" and just use 0x00800000 away from the current 24-bit timer value -- and the compare event does not fire as the time is not reached (the schedule or eventis adjusted again. 

The fix is trivial: just modify `ble_ll_rfmgmt.h` to return the current time as when the RF will be available:

```c
static inline uint32_t ble_ll_rfmgmt_enable_now(void) { return os_cputime_get32(); }
```

This fix is in pull request #688.  (Alternatively, this implementation could be moved to `ble_ll_rfmgmt.c`).

I will see about raising this with the upstream NimBLE package.

**Comment by heinrich-ulbricht** [comment]
Just updated to InfiniTime 1.5.0 and checked if the watch is now able to reconnect after switching the phone into flight mode and back. Unfortunately not. This specific problem persists.

So when switching the phone to flight mode and back the connection cannot be established (even after having connected watch and phone just minutes before) and the watch needs to be reset. But I'm really happy reading the findings in the comments above which I hope might help alleviate this problem.

**Comment by LlamasJM** [comment]
InfiniTime 1.5.0. Gadgetbridge. 
The watch is unable to reconnect to Gadgetbridge when I go out of range and return. I must go to Gadgetbridge and connect manually (the app is "Not connected" or "Waiting to reconnect"). 
And 18 hours after the update to 1.5.0., persistent issue: the watch is unable to reconnect to Gadgetbridge, and I had to reset it.

**Comment by bearclaws8** [comment]
@LlamasJM  I too have had a similar behavior since upgrading to 1.5.0. After the 18+ hour disconnect only a watch and phone restart allowed me to reconnect. Before PT 1.5.0 and GB 0.59.3 just restarting PT was enough.

The watch and GB were also frequently disconnecting during the day with me having to manually reconnect via GB. The link rarely automatically reconnected, causing missed notifications. I think this may have been on the GB end because sometimes just waking up my phone screen (Android 10) made the two reconnect. I'm not sure if this is new due to 1.5.0 or the GB 0.59.3 update. I will try to reload 1.4.0 later to check.

**Comment by ildar** [comment]
I see that GB works much worse then other BLE apps. NRF Connect sees the
gadget instantly, BLExplorer also. Linux PC also.
anyone seen anything suspicious in GB logs?

**Comment by bearclaws8** [comment]
I reinstalled 1.4.0 and have not had any connection/reconnecting issues (beyond having to reboot every 18+/- hours) with GB 0.59.3. I suspect that Bluetooth changes made for 1.5.0 may be causing this issue. Does anyone have advertising/broadcast data to help confirm? Does anyone see similar issues with nRF Connect? (I haven't tested that yet)

**Comment by LlamasJM** [comment]
Voilà! 1.6.0. has arrived, it seems that resolved the 18 hours issue. 
And now my watch reconnect with GB when I  go out and return near the phone. I think the solution was that: I removed GB from phone, and reinstalled it. And from this moment, PineTime reconnects automatically near the phone. At least, until this moment I am writing...

**Comment by ajostergaard** [comment]
There's still an issue when switching to airplane mode and back... I need to add a new device, select my current device from the list but then sneakily select 'do not pair' in the dialog. :) Seems likely it's a problem in GB though.

**Comment by heinrich-ulbricht** [comment]
Unfortunately yes, the airplane mode problem persists. I can only dream of reaching a connection that lasts 18 hours :/

**Comment by LlamasJM** [comment]
This morning, the PT disconnected from GB. Why? I don't know: probably at the moment when I connected the Heart Rate BPM, but I don't know exactly. 
It was impossible reconnected. I had to reset the watch. 
Now is working well. But some bugs persist...

**Comment by kennysc** [comment]
On my side I don't have this issue, I'm using:
Infinitime 1.6.0
Gadgetbridge 0.60.0
Android 11 on a Samsung S10

maybe Gadgetbridge 0.60.0 fixes this issue ?

**Comment by razzeee** [comment]
It's way better then before. It still disconnects from time to time, but so far I could always reconnect. Previously I needed to restart the pineTime for that to work.

**Comment by ajostergaard** [comment]
> On my side I don't have this issue, I'm using: Infinitime 1.6.0 Gadgetbridge 0.60.0 Android 11 on a Samsung S10
> 
> maybe Gadgetbridge 0.60.0 fixes this issue ?

Not fixed for me, still have to trick GB into connecting by adding and then selecting "don't add" but ONLY after airplane mode.

Same versions except a Fairphone.

**Comment by bearclaws8** [comment]
I've been running 1.6.0 for almost 36 hours now and have not had to restart the PT once to fix it advertising/showing up.

With GB 0.59.3 or 0.60 I have issues reconnecting after anything that cycles Bluetooth off, including toggling BT, toggling airplane mode, or sometimes a restart. For GB I usually have to remove the device to re-add it after restarting the phone. Sometimes it takes several phone restarts, sometimes one, or sometimes just letting them sit disconnected for minutes/hours. Even clearing the GB cache does not help.

With nRF Connect 4.24.3 I also have issues reconnecting after a BT toggle. However, a restart allows it to reconnect. IF GB has failed to connect before trying nRF then I cannot get nRF to connect to the PT. Connecting nRF before GB does not always allow GB to reconnect to the PT.

I even tried to fix this by restarting my PT and that did not change the behavior noted above. So I _think_ (maybe??) that the PT side of this issue may be resolved. Unless PT is storing BLE connection info across restarts that disables new connections, which I doubt. Do we think that this may be closed from a PT perspective but still open from a GB, nRF Connect (maybe Android OS) perspective?

**Comment by ainola** [comment]
I think it is also worth mentioning the incredible work done by those that have been looking into these issues. This is very clearly a difficult bug and the amount of time and pain to fix them, let alone track them down, is incredible.

**Comment by lordfeck** [comment]
I wanted to add my congratulations and thanks for this fix. I've had my PineTime for a month and this has been one of the few (minor) troubles with InfinniTime. Now that it's fixed, this humble watch moves closer to rivalling the big boys!

**Comment by stillmanff** [comment]
I've just spent a half hour restarting the PineTime, restarting GadgetBridge, and trying to pair the two through the BT control panel in Android. Nothing works. The closest I get is a failed "Pairing..." message in the control panel, and "Creating bond with InfiniTime"/ "Bonding with Infinitime"/"Attempting to pair with InfiniTime"/"Trying to connect with InfiniTime" in GadgetBridge, all of which end with the watch and phone not connected. Doesn't matter if the screen is on or off once GadgetBridge or the OS discovers the watch. I can seen that things are much improved in 1.5 and 1.6 but I'm running 1.4 and can't upgrade until I can connect. Any advice?

**Comment by PhilDevProg** [comment]
> I've just spent a half hour restarting the PineTime, restarting GadgetBridge, and trying to pair the two through the BT control panel in Android. Nothing works. The closest I get is a failed "Pairing..." message in the control panel, and "Creating bond with InfiniTime"/ "Bonding with Infinitime"/"Attempting to pair with InfiniTime"/"Trying to connect with InfiniTime" in GadgetBridge, all of which end with the watch and phone not connected. Doesn't matter if the screen is on or off once GadgetBridge or the OS discovers the watch. I can seen that things are much improved in 1.5 and 1.6 but I'm running 1.4 and can't upgrade until I can connect. Any advice?

You could try another app (e.g. NRF Connect).

**Comment by tmilburn** [comment]
@stillmanff InfiniTime doesn't require pairing/bonding so don't try that through the bluetooth control panel. Unpair it from the bluetooth control panel and just open gadgetbridge and try connecting using that.

**Comment by stillmanff** [comment]
It didn't successfully pair at all, so there's nothing to unpair. However, I updated Gadgetbridge to 0.60, an update that didn't show up last night when I was trying. As soon as I did that, the PineTime connected, first try, with pairing. Thanks for the advice.

**Comment by GithubUser5462** [comment]
It's still borked for me on the latest GB and IT versions but i have found a bit of a workaround.
I'm posting it because i'm curious if it works for anyone else + it might help identifying the issue.

Basically, when GB gets stuck on "connecting" you press and hold the tile until it says "Not connected.".
Then you enable location services (GPS) and go to "Device discovery" (the + button).
Press the physical button on the watch so it wakes up and then press the "Start discovery" / "Stop scanning" button multiple times. Keep pressing the button until the watch shows up, it should show up quickly, after 2 to 5 presses in my case.
Stop scanning, but don't pair the watch, go back and try to reconnect, it should successfully reconnect now.

I know that it is not much of a workaround, but it doesn't require rebooting or re-pairing.
I hope that this new info and it sheds some light on the issue.

**Comment by bearclaws8** [comment]
@akutuzov  I started having the issue again without using the HRM and within a day of reconnecting to GB via phone restart. So I'm not sure it is explicit HRM related, maybe it is related to #302.

**Comment by infinitytec** [comment]
I have trouble when my phone restarts or the watch goes out of range.

The range issue may be better since 1.6 though.

**Comment by JackofSpades707** [comment]
I went ahead and updated the firmware and it seems like the majority of the issues have subsided if not been entirely addressed

**Comment by Avamander** [comment]
Each and all issues related to reconnection or pairing should now be held until https://github.com/InfiniTimeOrg/InfiniTime/pull/796 is merged. Then a new issue must be opened for any new issues.
